BFWControls/Modules/Layout/View/PileView.swift

```diff
@@ -3,57 +3,117 @@
     open override var intrinsicContentSize: CGSize {
-        return CGSize(width: UIViewNoIntrinsicMetric,
-                      height: isHorizontal
-                        ? subviewHeight
-                        : subviewCount * subviewHeight + (subviewCount - 1) * gapHeight)
+        var width: CGFloat = 0
+        var height: CGFloat = 0
+        if isHorizontal {
+            width  = subviews.reduce(0) { (result, subview) in
+                return (result + subview.intrinsicContentSize.width)
+            }
+            width += (subviewCount - 1)*gapWidth
+            height  = subviews.reduce(0) { (result, subview) in
+                return max(result, subview.intrinsicContentSize.height)
+            }
+        } else  {
+            width  = subviews.reduce(0) { (result, subview) in
+                return max(result, subview.intrinsicContentSize.width)
+            }
+            height  = subviews.reduce(0) { (result, subview) in
+                return (result + subview.intrinsicContentSize.height)
+            }
+            height += (subviewCount - 1)*gapHeight
+        }
+        return CGSize(width: width, height: height)
+    }
+    
+    
+    /// Returns origin for a subview
+    ///
+    /// - Parameters:
+    ///   - subview: Current subview
+    ///   - previousSubview: Previous subview
+    /// - Returns: Origin as CGPoint
+    private func getOrigin(for subview: UIView, previousSubview: UIView?) -> CGPoint {
+        var subviewOriginY: CGFloat = 0.0
+        var subviewOriginX: CGFloat = 0.0
+        if isHorizontal {
+            subviewOriginX = previousSubview != nil
+                ? previousSubview!.frame.origin.x + previousSubview!.bounds.size.width + gapWidth
+                : 0.0
+        } else  {
+            subviewOriginY = previousSubview != nil
+                ? previousSubview!.frame.origin.y + previousSubview!.bounds.size.height + gapHeight
+                : 0.0
+        }
+        return CGPoint(x: subviewOriginX, y: subviewOriginY)
     }
     
+    /// Subviews layed out horizontally or vertically.
     open override func layoutSubviews() {
         invalidateIntrinsicContentSize()
         for (index, subview) in subviews.enumerated() {
-            let origin = isHorizontal
-                ? CGPoint(x: (subviewWidth + gapWidth) * CGFloat(index),
-                          y: 0.0)
-                : CGPoint(x: 0.0,
-                          y: (subviewHeight + gapHeight) * CGFloat(index))
+            for eachContraint in subview.constraints {
+                subview.removeConstraint(eachContraint)
+            }
+            let subviewWidth = min(subview.intrinsicContentSize.width, bounds.size.width)
+            let subviewHeight = min(subview.intrinsicContentSize.height, bounds.size.height)
+            let origin = getOrigin(for: subview, previousSubview: (index > 0 ? subviews[index-1] : nil))
```

Comment:
Instead of calculating whether to send a nil, which the function then re-interprets, perhaps it would be simpler to do it all here, without a function, eg:

```swift
let origin: CGPoint
if index == 0 {
    origin = CGPoint.zero
} else {
    let previousSubview = subviews[index - 1]
    origin = isHorizontal
        ? CGPoint(x: previousSubview.frame.maxX + gapWidth, 0)
        : CGPoint(x: 0, y: previousSubview.frame.maxY + gapHeight)
}
```